File: ios/Source/DTOs/Web3RequestDTO.swift
// Copyright (c) 2018-2019 Coinbase, Inc. <https://coinbase.com/>
// Licensed under the Apache License, version 2.0

import CBCore

struct Web3RequestDTO<T: Codable>: Codable, JSONDeserializable {
    let type: String = "WEB3_REQUEST"
    let id: String
    let origin: URL
    let request: Web3Request<T>
}

struct Web3Request<T: Codable>: Codable {
    let method: RequestMethod
    let params: T
}

struct RequestEthereumAccountsParams: Codable {
    let appName: String?
    let appLogoUrl: URL?
}

struct SignEthereumMessageParams: Codable {
    let message: String
    let address: String
    let addPrefix: Bool
    let typedDataJson: String?
}

struct SignEthereumTransactionParams: Codable {
    let fromAddress: String
    let toAddress: String?
    let weiValue: String
    let data: String
    let nonce: Int?
    let gasPriceInWei: String?
    let gasLimit: String?
    let chainId: Int
    let shouldSubmit: Bool
}

struct SubmitEthereumTransactionParams: Codable {
    let signedTransaction: String
    let chainId: Int
}

File: android/walletlink/src/main/java/com/coinbase/walletlink/dtos/Web3RequestDTO.kt
package com.coinbase.walletlink.dtos

import com.coinbase.wallet.core.util.JSON
import com.coinbase.walletlink.models.RequestMethod
import com.squareup.moshi.Types
import java.net.URL

internal data class Web3RequestDTO<T>(
    val id: String,
    val origin: URL,
    val request: Web3Request<T>
) {
    companion object
}

internal inline fun <reified T> Web3RequestDTO.Companion.fromJsonString(json: String): Web3RequestDTO<T>? {
    val type = Types.newParameterizedType(Web3RequestDTO::class.java, T::class.java)
    val adapter = JSON.moshi.adapter<Web3RequestDTO<T>>(type)

    return adapter.fromJson(json)
}

internal inline fun <reified T> Web3RequestDTO.Companion.fromJson(json: ByteArray): Web3RequestDTO<T>? {
    val jsonString = String(json, Charsets.UTF_8)
    return fromJsonString(jsonString)
}

internal data class Web3Request<T>(val method: RequestMethod, val params: T)

internal data class RequestEthereumAccountsParams(val appName: String, val appLogoUrl: URL?)

internal data class SignEthereumMessageParams(val message: String, val address: String, val addPrefix: Boolean)

internal data class SignEthereumTransactionParams(
    val fromAddress: String,
    val toAddress: String?,
    val weiValue: String,
    val data: String,
    val nonce: Int?,
    val gasPriceInWei: String?,
    val gasLimit: String?,
    val chainId: Int,
    val shouldSubmit: Boolean,
    val typedDataJson: String?
)

internal data class SubmitEthereumTransactionParams(val signedTransaction: String, val chainId: Int)
